File: advent-2023/python/q1.py
import re


def part1(file):
    nums = [[ch for ch in line if ch.isdigit()]
            for line in open(file).read().split('\n')]
    total = sum([int(num[0] + num[-1]) for num in nums])
    print(total)


num_dt = {"one": 1, "two": 2, "three": 3, "four": 4,
          "five": 5, "six": 6, "seven": 7, "eight": 8, "nine": 9}
"three eight"
"eight three"
"five eight"
""


def gen_overlaps():
    overlaps = []
    for num_s in num_dt.keys():
        for other_num in num_dt.keys():
            if num_s[-1] == other_num[0]:
                overlaps.append((num_s + other_num, num_s + other_num[1:]))
    return overlaps


def replace_line(line, overlaps):
    for overlap in overlaps:
        des, src = overlap
        line = line.replace(src, des)
    return line


def part2(file):
    pattern = r"one|two|three|four|five|six|seven|eight|nine|[1-9]"
    overlaps = gen_overlaps()
    searches = [re.findall(pattern, replace_line(line, overlaps))
                for line in open(file).read().split('\n')]
    total = 0
    for s in searches:
        # print(s)
        first = num_dt[s[0]] if not s[0].isdigit() else s[0]
        last = num_dt[s[-1]] if not s[-1].isdigit() else s[-1]
        total += int(str(first) + str(last))
    print(total)


if __name__ == "__main__":
    file = "./inputs/q1.txt"
    # part1(file)
    part2(file)
    # use start with

File: advent-2023/c/program/q1.c
#define _CRT_SECURE_NO_WARNINGS

#include "call_lib.h"

bool starts_with(char line[], char new_num[]) {
    size_t i;

    for(i=0; new_num[i] != '\0'; i++) {
        if (*(line + i) != *(new_num + i)) {
            return FALSE;
        }
    }
    
    return TRUE;
}

void part1(char* input) {
    char new_nums[9][6] = {"one","two","three","four","five","six","seven","eight","nine"};

    char line[50];
    int result = 0;
    int result_p2 = 0;
    int read = 0, offset = 0;
    while (sscanf(input + offset, "%s%n", line, &read) != EOF) {
        size_t i, j ;
        char* list = create_list(char);
        char* list_p2 = create_list(char);
        for (i=0; line[i] != '\0'; i++) {
            if ('1' <= line[i] && line[i] <= '9') {
                push_list(list, line[i]);
                push_list(list_p2, line[i]);
            } else {
                for (j=0; j < 9; j++) {
                    if (starts_with(line + i, new_nums[j])) {
                        char num = j + 1 + '0';
                        push_list(list_p2, num);
                        break;
                    }
                }
            }
        }
        result += (list[0] - '0') * 10 + (list[list_len(list) - 1] - '0');
        result_p2 += (list_p2[0] - '0') * 10 + (list_p2[list_len(list_p2) - 1] - '0');
        destroy_list(list);
        destroy_list(list_p2);
        offset += read;
    }
    printf("%d\n", result);
    printf("%d\n", result_p2);
}

void part2(char* input) {
}

SOLUTION("./inputs/q1.txt")

